Move MediaStreamTrack capture state handling to Document

youennf opened the pull request:
#### 31557cfb95e2fbd2e27f4ca6de470b2b91bf56e1
<pre>
Move MediaStreamTrack capture state handling to Document
<a href="https://bugs.webkit.org/show_bug.cgi?id=267325">https://bugs.webkit.org/show_bug.cgi?id=267325</a>
<a href="https://rdar.apple.com/120776008">rdar://120776008</a>

Reviewed by Jean-Yves Avenard and Eric Carlson.

MediaStreamTrack may be transferred from a document to a worker.
As per spec, transferring a track does not transfer the ownership of source.
This ensures UA capture states remain the same even if a track gets transferred.

To align with this and prepare track transfer, we are moving capture state handling to Document.
A document will own directly RealtimeMediaSource instead of going through MediaStreamTrack, since a transferred MediaStreamTrack gets closed.
We introduce the possibility for a Document to register itself as an owner of a RealtimeMediaSource.
The RealtimeMediaSource will notify the Document when it either starts or ends.
The Document updates its capture state based on this signal.

When the document goes away, it will stop all sources like would have been done previously with any remaining MediaStreamTrack.

We continue updating code to limit access to a MediaStreamTrack source.

* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::create):
(WebCore::MediaStreamTrack::MediaStreamTrack):
(WebCore::MediaStreamTrack::~MediaStreamTrack):
(WebCore::trackCaptureState):
(WebCore::MediaStreamTrack::mediaState const):
(WebCore::MediaStreamTrack::trackEnded):
(WebCore::MediaStreamTrack::trackMutedChanged):
(): Deleted.
(WebCore::sourceCaptureState): Deleted.
(WebCore::MediaStreamTrack::captureState): Deleted.
(WebCore::MediaStreamTrack::updateCaptureAccordingToMutedState): Deleted.
(WebCore::MediaStreamTrack::updateVideoCaptureAccordingMicrophoneInterruption): Deleted.
(WebCore::MediaStreamTrack::updateToPageMutedState): Deleted.
(WebCore::trackTypeForMediaProducerCaptureKind): Deleted.
(WebCore::MediaStreamTrack::endCapture): Deleted.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
(WebCore::MediaStreamTrack::isDetached const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
(WebCore::Document::visibilityStateChanged):
(WebCore::captureSourceState):
(WebCore::Document::computeCaptureState const):
(WebCore::Document::updateIsPlayingMedia):
(WebCore::Document::pageMutedStateDidChange):
(WebCore::updateCaptureSourceToPageMutedState):
(WebCore::Document::addCaptureSource):
(WebCore::Document::captureSourceStateChanged):
(WebCore::Document::updateCaptureAccordingToMutedState):
(WebCore::trackTypeForMediaProducerCaptureKind):
(WebCore::Document::stopMediaCapture):
(WebCore::Document::updateVideoCaptureAccordingMicrophoneInterruption):
* Source/WebCore/dom/Document.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::start):
(WebCore::RealtimeMediaSource::end):
(WebCore::RealtimeMediaSource::registerOwnerCallback):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.cpp:
(WebKit::RemoteRealtimeVideoSource::clone):

Canonical link: <a href="https://commits.webkit.org/273218@main">https://commits.webkit.org/273218@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0586aac3e87c0a395c4edeeda049114b150e7125

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13582 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37447 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31370 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35903 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15986 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10667 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30317 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35288 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11498 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10050 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10112 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31029 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38712 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31557 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31352 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36161 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10229 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34135 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12057 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7966 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10776 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11120 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
```diff
@@ -426,12 +429,22 @@ void RealtimeMediaSource::end(Observer* callingObserver)
+{
+    ASSERT(isMainThread());
+    if (!m_registerOwnerCallback)
+        m_registerOwnerCallback = WTFMove(callback);
```
Should this do something like `ASSERT(!m_registerOwnerCallback || m_registerOwnerCallback == callback)`?

youennf replied:
Will add the ASSERT, registration of a source should only happen once.

eric-carlson commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8907,7 +9067,24 @@ void Document::mediaStreamCaptureStateChanged()
 }
 
-#endif
+void Document::updateVideoCaptureAccordingMicrophoneInterruption(bool isMicrophoneInterrupted)
```
`...According...` isn't quite correct. Maybe either `...AccordingTo...` or `updateVideoCaptureStateForMicrophoneInterruptionState`?

youennf replied:
Will change to `updateVideoCaptureStateForMicrophoneInterruption`

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -65,23 +65,15 @@
 
-    if (track->isCaptureTrack())
-        track->updateToPageMutedState();
+    if (track->isCaptureTrack() && !track->ended())
```
```
if (RefPtr document = dynamicDowncast<Document>(context); track->isCaptureTrack() && !track->ended() && document)
  document->addCaptureSource...
```

would be nicer I believe

youennf replied:
A track can be a capture track only if its context is a Document, no need for a check here.

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -486,113 +463,16 @@ MediaProducerMediaStateFlags sourceCaptureState(RealtimeMediaSource& source)
-{
-    ASSERT(isCaptureTrack());
     RefPtr context = scriptExecutionContext();
+    if (!context || !is<Document>(context) || !downcast<Document>(context)->page())
```
`if (RefPtr document = dynamicDowncast<Document>(context)....`

youennf replied:
Will change

jyavenard commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4965,6 +4971,66 @@ void Document::noteUserInteractionWithMediaElement()
+    case CaptureDevice::DeviceType::Unknown:
+        ASSERT_NOT_REACHED();
+        break;
+    }
```
that seems like a lot of duplicated code, couldn't we make the static method take a `CaptureDevice::DeviceType` instead and put this code in common?

youennf replied:
Will check what I can do there.

youennf replied:
Done by sharing code with MediaStreamTrack.

jyavenard commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8907,7 +9067,24 @@ void Document::mediaStreamCaptureStateChanged()
+    if (!page)
+        return;
+
+    for (auto& document : allDocuments()) {
```
shouldn't we take a strong reference here just in case? seems to have been a generic trend lately

youennf replied:
I'll add it.

youennf committed:
Move MediaStreamTrack capture state handling to Document
https://bugs.webkit.org/show_bug.cgi?id=267325
rdar://120776008

Reviewed by Jean-Yves Avenard and Eric Carlson.

MediaStreamTrack may be transferred from a document to a worker.
As per spec, transferring a track does not transfer the ownership of source.
This ensures UA capture states remain the same even if a track gets transferred.

To align with this and prepare track transfer, we are moving capture state handling to Document.
A document will own directly RealtimeMediaSource instead of going through MediaStreamTrack, since a transferred MediaStreamTrack gets closed.
We introduce the possibility for a Document to register itself as an owner of a RealtimeMediaSource.
The RealtimeMediaSource will notify the Document when it either starts or ends.
The Document updates its capture state based on this signal.

When the document goes away, it will stop all sources like would have been done previously with any remaining MediaStreamTrack.

We continue updating code to limit access to a MediaStreamTrack source.

* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::create):
(WebCore::MediaStreamTrack::MediaStreamTrack):
(WebCore::MediaStreamTrack::~MediaStreamTrack):
(WebCore::trackCaptureState):
(WebCore::MediaStreamTrack::mediaState const):
(WebCore::MediaStreamTrack::trackEnded):
(WebCore::MediaStreamTrack::trackMutedChanged):
(): Deleted.
(WebCore::sourceCaptureState): Deleted.
(WebCore::MediaStreamTrack::captureState): Deleted.
(WebCore::MediaStreamTrack::updateCaptureAccordingToMutedState): Deleted.
(WebCore::MediaStreamTrack::updateVideoCaptureAccordingMicrophoneInterruption): Deleted.
(WebCore::MediaStreamTrack::updateToPageMutedState): Deleted.
(WebCore::trackTypeForMediaProducerCaptureKind): Deleted.
(WebCore::MediaStreamTrack::endCapture): Deleted.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
(WebCore::MediaStreamTrack::isDetached const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
(WebCore::Document::visibilityStateChanged):
(WebCore::captureSourceState):
(WebCore::Document::computeCaptureState const):
(WebCore::Document::updateIsPlayingMedia):
(WebCore::Document::pageMutedStateDidChange):
(WebCore::updateCaptureSourceToPageMutedState):
(WebCore::Document::addCaptureSource):
(WebCore::Document::captureSourceStateChanged):
(WebCore::Document::updateCaptureAccordingToMutedState):
(WebCore::trackTypeForMediaProducerCaptureKind):
(WebCore::Document::stopMediaCapture):
(WebCore::Document::updateVideoCaptureAccordingMicrophoneInterruption):
* Source/WebCore/dom/Document.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::start):
(WebCore::RealtimeMediaSource::end):
(WebCore::RealtimeMediaSource::registerOwnerCallback):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.cpp:
(WebKit::RemoteRealtimeVideoSource::clone):

Canonical link: https://commits.webkit.org/273218@main
